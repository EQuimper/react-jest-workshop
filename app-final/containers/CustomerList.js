/**
 * KENT: DON'T FORGET TO UPDATE THIS TO NOT
 * USE PROPS.STORE EVERYWHERE BEFORE YOUR
 * MIDWESTJS TALK :)
 */
import React, {PropTypes, Component} from 'react'
import store from '../store/Customers'

class CustomerList extends Component {
  constructor(props) {
    super(props)
    this.state = {
      customers: props.store.getCustomers(),
    }
  }
  componentDidMount() {
    this.unsubscribe = this.props.store.subscribe(this.updateStateWithCustomers)
  }
  componentWillUnmount() {
    this.unsubscribe()
  }

  updateStateWithCustomers = () => {
    const customers = this.props.store.getCustomers()
    this.setState({customers})
  };

  render() {
    const {customers} = this.state
    if (customers.length === 0) {
      return <NoCustomers />
    } else {
      return <ListOfCustomers customers={customers} />
    }
  }
}

CustomerList.defaultProps = {store}

CustomerList.propTypes = {
  store: PropTypes.shape({
    getCustomers: PropTypes.func,
    subscribe: PropTypes.func,
  }),
}

function ListOfCustomers({customers}) {
  return (
    <div>
      Here is your list of customers!
      <ul>
        {customers.map((c, i) => <li key={i}>{c.name}</li>)}
      </ul>
    </div>
  )
}

ListOfCustomers.propTypes = {
  customers: PropTypes.array,
}

function NoCustomers() {
  return (
    <div>
      <div className="wrapper-that-needs-to-happen-for-some-reason">
        You have absolutely no customers. Better get to work!
      </div>
    </div>
  )
}

export default CustomerList
