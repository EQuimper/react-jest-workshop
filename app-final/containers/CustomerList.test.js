import React from 'react'
import renderer from 'react-test-renderer'
import getStoreStub from '../store/Customers.stub'

import CustomerList from './CustomerList'

test('should render no customers', () => {
  const component = renderCustomerList()
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('should render customers', () => {
  const {store} = getStoreStub([{name: 'Bob'}, {name: 'Joanna'}])
  const component = renderCustomerList({store})
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('should respond to store updates', () => {
  const {store, updateCustomers} = getStoreStub()
  const component = renderCustomerList({store})
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
  updateCustomers([{name: 'Jill'}, {name: 'Fred'}])
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

/**
 * Render the <CustomerList /> component with enzyme and return the wrapper
 * @param {Object} props={} - the props to apply to the <CustomerList /> component
 * @return {Wrapper} - the enzyme wrapper
 */
function renderCustomerList(props = {}) {
  return renderer.create(<CustomerList {...getProps(props)} />)
}

/**
 * Get the default props for tests
 * @param {Object} props={} the props to apply to the <CustomerList /> component
 * @return {Object} store, actions, and the props you've passed (overwriting the store and actions)
 */
function getProps(props = {}) {
  const {store} = getStoreStub()
  return {store, ...props}
}
