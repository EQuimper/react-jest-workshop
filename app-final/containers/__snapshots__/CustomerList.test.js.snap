exports[`test should render customers 1`] = `
<div>
  Here is your list of customers!
  <ul>
    <li>
      Bob
    </li>
    <li>
      Joanna
    </li>
  </ul>
</div>
`;

exports[`test should render no customers 1`] = `
<div>
  <div
    className="wrapper-that-needs-to-happen-for-some-reason">
    You have absolutely no customers. Better get to work!
  </div>
</div>
`;

exports[`test should respond to store updates 1`] = `
<div>
  <div
    className="wrapper-that-needs-to-happen-for-some-reason">
    You have absolutely no customers. Better get to work!
  </div>
</div>
`;

exports[`test should respond to store updates 2`] = `
<div>
  Here is your list of customers!
  <ul>
    <li>
      Jill
    </li>
    <li>
      Fred
    </li>
  </ul>
</div>
`;
